src/Collector.cs

```diff
@@ -79,25 +83,36 @@ private void StartBackgroundWorker()
 
     private void WorkerOnDoWork(object sender, DoWorkEventArgs doWorkEventArgs)
     {
+      var failures = 0;
+      Socket socket = null;
+
+      // This is not really the worker loop, it is just very careful connect/authenticate
+      // If you want the worker loop, look at SendQueuedMessages
       while (!_worker.CancellationPending)
         {
-          Socket socket = null;
-          var failures = 0;
-
           try
             {
               socket = Connect();
-              Authenticate(socket);
-              failures = 0;
+              if(!Authenticate(socket))
+                {
+                  socket.Close();
+                  // Throttle calls to authenticate so that you do not flood the buffer if collector is down/slow for a few seconds.
+                  // Ultimately should share logic with the backoff stuff below, but did not want to just throw here.
+                  Thread.Sleep(5000);
+                  continue;
+                }
               SendQueuedMessages(socket);
+              failures = 0;
+              socket.Close();
```

Comment:
No other agents do a close here, but I think it's a nice move for handling an exiting process which may cause `SendQueuesMessages` to return.

If we're going to do it, I think we should call `socket.shutdown(SocketShutdown.Both)` here, just like the code on 115.

From https://msdn.microsoft.com/en-us/library/system.net.sockets.socket.shutdown(v=vs.110).aspx
> When using a connection-oriented Socket, always call the Shutdown method before closing the Socket. This ensures that all data is sent and received on the connected socket before it is closed.